source/utf8/unchecked.h

```diff
@@ -226,8 +257,7 @@ namespace utf8
             octet_iterator base () const { return it; }
             uint32_t operator * () const
             {
-                octet_iterator temp = it;
-                return utf8::unchecked::next(temp);
+                return utf8::unchecked::peek(it);
```

Comment:
How about holding some state instead? E.g. on advance just read the next char into a local uint32. Need to handle first char and `--` somehow and comparison, but this could work...

Or: Can you check for the istream iterators? They got too have some tag to distinguish them... And then e.g. require a bidirectional iterator.